src/main/java/com/google/devtools/build/lib/skyframe/SkyframeExecutor.java

```diff
@@ -984,6 +984,37 @@ private static boolean isAnalysisPhaseActionLookupKey(SkyKey key) {
     return key instanceof ActionLookupKey && !(key instanceof ActionTemplateExpansionKey);
   }
 
+  /**
+   * A variant of {@link #DEFAULT_EVENT_FILTER_WITH_ACTIONS} that also handles Bzlmod events, which
+   * are only relevant for Bazel and its support for external repositories.
+   */
+  public static final EventFilter DEFAULT_EVENT_FILTER_WITH_ACTIONS_AND_EXTERNAL_REPOS =
+      new EventFilter() {
+        @Override
+        public boolean storeEvents() {
+          return true;
+        }
+
+        @Override
+        public boolean shouldPropagate(SkyKey depKey, SkyKey primaryKey) {
+          // Bzlmod events do not need to be propagated to analysis nodes. They are propagated to
+          // the node for the main repo mapping, which is unconditionally requested in a separate
+          // Skyframe evaluation for every build, thus ensuring correct incremental behavior.
+          if (isBzlmodKey(depKey)) {
+            return isBzlmodKey(primaryKey);
+          }
+          return DEFAULT_EVENT_FILTER_WITH_ACTIONS.shouldPropagate(depKey, primaryKey);
+        }
+      };
+
+  private static boolean isBzlmodKey(SkyKey key) {
+    // The return value of Class#getPackageName() and string literals are always interned.
+    // https://github.com/openjdk/jdk/blob/3e185c70feef3febf75c58a5d4d394a4b772105f/src/java.base/share/classes/java/lang/Class.java#L1258
+    return key.getClass().getPackageName() == "com.google.devtools.build.lib.bazel.bzlmod"
```

Comment:
What's the deal with this package check?

Reply:
I replied on https://github.com/bazelbuild/bazel/pull/22058#discussion_r1576562144